File: src/patterns/creational/singleton/carstorage.nim
type
    CarStorage* = ref object
        cars: seq[string]


proc newCarStorage(): CarStorage =
    return CarStorage()


var instance = newCarStorage()


proc getInstance*(): CarStorage =
    return instance


proc getCars*(self: CarStorage): seq[string] =
    return self.cars

proc addCar*(self: CarStorage, car: string) =
    self.cars.add(car)

File: src/patterns/creational/singleton/carstorage.go
package singleton


type CarStorage struct {
	cars []string
}


func newCarStorage() *CarStorage {
	return &CarStorage{
		cars: []string{},
	}
}


var instance *CarStorage


func GetInstance() *CarStorage {
	if instance == nil {
		instance = newCarStorage()
		return instance
	}
	return instance
}

func (self *CarStorage) GetCars() []string {
	return self.cars
}

func (self *CarStorage) AddCar(car string) {
	self.cars = append(self.cars, car)
}